Make it possible to specify output directory.

sdomme opened the pull request:
As user I would like to specify an output directory, so I can control whereto snakepit will generate the spec-file.

sdomme committed:
make it possible to specify output directory.

esc commented on src/main/python/snakepit/main.py:
```diff
@@ -56,6 +57,13 @@ class TemplateNoteFoundException(Exception):
     pass
 
 
+class OutputDirectoryNotWritable(Exception):
```
Rather than checking first, you should probably `try` `except` when attempting to write.

esc replied:
Also, if you write code to check this, you should probably test it as part of the cram tests.

esc commented on src/main/scripts/snakepit:
```diff
@@ -4,16 +4,17 @@
+  --debug               Enable debug output.
+  --build=<build>       The build number [default: 0]
+  -f, --force           Force overwrite of output.
+  --output=<direcotry>  output directory for SPEC-File
```
Please start the description with a capital letter and end it with a period.

esc replied:
Question: what if I want to specify the spec-file name too?

esc commented on src/main/python/snakepit/main.py:
```diff
@@ -80,6 +88,14 @@ def get_pypi_metadata(package, url='https://pypi.python.org/pypi/'):
 
 
+def custom_output_filename(filename, output_directory):
+    output_filename = "{0}/{1}".format(output_directory, filename)
```
Please use `os.path.join` rather than `/`. `os.path` is imported as `osp` in this module.

esc commented:
@sdomme this looks really good! Please fixup the minor nitpicks and I'll merge and release. (no automatic releases as of yet.

esc commented:
Also, remember to delete the TODO entry in the README as part of the pull-request.

sdomme committed:
Deleted the output TODO

sdomme committed:
Repair description in docstring for output

sdomme committed:
Using os.path.join makes it OS-independent.

sdomme committed:
Let's make it less complicated. "with open"-should handle the file exceptions.

sdomme committed:
fix unittest

sdomme committed:
test if the output directory is writable

sdomme committed:
User can now specify the output as:
- any filename
- any directory
- any directory + any filename
assuming it is writable. The code decide how to handle this.

sdomme commented:
@esc I think I implemented all of your suggestions. ;-)

esc commented on src/main/python/snakepit/main.py:
```diff
@@ -80,6 +80,10 @@ def get_pypi_metadata(package, url='https://pypi.python.org/pypi/'):
 
 
+def custom_output_filename(filename, output_directory):
+    return "{0}".format(osp.join(output_directory, filename))
```
no need to format here, just join.

esc commented on src/main/python/snakepit/main.py:
```diff
@@ -120,7 +124,14 @@ def main(arguments):
     # get the output filename
-    output_filename = default_output_filename(yaml_spec)
+    if arguments['--output']:
+        if osp.isdir(arguments['--output']):
```
This is a nice piece of logic.

esc commented on src/cmdlinetest/test_snakepit.t:
```diff
@@ -88,6 +91,29 @@
    '<file>': 'snakepit.yaml'}
   Writing output to: 'moto.spec'
+
+# test specified output is an existent directory
```
The negative case testing is good, but you should test the positive ones too. i.e.:

```
$snakepit snakepit.yaml --output foo.yaml
$ ls foo.yaml
foo.yaml
```

and 

```
$ mkdir bar
$ snakepit snakepit.yaml --output bar/foo.yaml
$ ls bar/
foo.yaml

```

Or something similar.

esc commented:
@sdomme sweet! Looks much better, but a few things left to go.

esc commented:
@sdomme also I've given the company access to the project on PyPi, feel free to setup CLD like in the pybuilder_aws_lambda_plugin (hint: you'll need the pybuilder setup.py)

sdomme committed:
Just join.

sdomme committed:
Some new cram tests to check if the logic for file creation is working

sdomme commented:
Done. Please review

esc commented:
Perfect!
